tsk/util/Bitlocker/BitlockerParser.cpp

```diff
@@ -0,0 +1,1553 @@
+    }
+
+    if (!m_haveRecoveryPassword && protectionType == BITLOCKER_KEY_PROTECTION_TYPE::RECOVERY_PASSWORD) {
+        writeError("BitlockerParser::parseVMKEntry: Can't process recovery password-protected VMK since we have no recovery password");
+        return BITLOCKER_STATUS::NEED_PASSWORD;
+    }
+
+    // The expectation is that we'll have a stretch key entry
+    list<MetadataValue*> stretchKeys;
+    getMetadataValues(vmkValue->getProperties(), BITLOCKER_METADATA_VALUE_TYPE::STRETCH_KEY, stretchKeys);
+    if (stretchKeys.empty()) {
+        writeError("BitlockerParser::parseVMKEntry: Volume Master Key had no stretch key entry");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    MetadataValueStretchKey* stretchKey = dynamic_cast<MetadataValueStretchKey*>(stretchKeys.front());
+    if (stretchKey == nullptr) {
+        writeError("BitlockerParser::parseVMKEntry: Error casting MetadataValueStretchKey");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Use password/recovery password to create intermediate stretched key
+    uint8_t stretchedKey[BITLOCKER_STRETCH_KEY_SHA256_LEN];
+    BITLOCKER_STATUS ret;
+    if (protectionType == BITLOCKER_KEY_PROTECTION_TYPE::PASSWORD) {
+        ret = stretchKey->parseStretchKeyUsingPassword((uint8_t*)m_passwordHash, SHA256_DIGEST_LENGTH, stretchedKey, BITLOCKER_STRETCH_KEY_SHA256_LEN);
+    }
+    else {
+        ret = stretchKey->parseStretchKeyUsingPassword((uint8_t*)m_recoveryPasswordHash, SHA256_DIGEST_LENGTH, stretchedKey, BITLOCKER_STRETCH_KEY_SHA256_LEN);
+    }
+    if (ret != BITLOCKER_STATUS::SUCCESS) {
+        writeError("BitlockerParser::parseVMKEntry: Error creating intermediate stretched key");
+        memset(stretchedKey, 0, BITLOCKER_STRETCH_KEY_SHA256_LEN);
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // There should also be one encrypted AES-CCM key entry
+    list<MetadataValue*> encryptedKeys;
+    getMetadataValues(vmkValue->getProperties(), BITLOCKER_METADATA_VALUE_TYPE::AES_CCM_ENCRYPTED_KEY, encryptedKeys);
+    if (encryptedKeys.empty()) {
+        writeError("BitlockerParser::parseVMKEntry: Volume Master Key had no encrypted key entry");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    MetadataValueAesCcmEncryptedKey* aesCcmKey = dynamic_cast<MetadataValueAesCcmEncryptedKey*>(encryptedKeys.front());
+    if (aesCcmKey == nullptr) {
+        writeError("BitlockerParser::parseVMKEntry: Error casting MetadataValueStretchKey");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Decrypt it using the stretched key, which should produce a MetadataEntry of type KEY.
+    // This includes testing a 16-byte message authentication code to verify that
+    // the decrypted key is correct.
+    MetadataEntry* keyEntry = nullptr;
+    ret = aesCcmKey->decrypt(stretchedKey, BITLOCKER_STRETCH_KEY_SHA256_LEN, &keyEntry);
+    if (ret != BITLOCKER_STATUS::SUCCESS) {
```

Comment:
should we add a comment that keyEntry will always be a valid allocated obj if bitlocker_Status is success since we aren't verifying that its not a nullptr. I thought we had an issue until I went back to see what decrypt() was doing.